src/Database/Oracle/Simple/Internal.hs

```diff
@@ -825,9 +831,31 @@ data DPIData = DPIData
   deriving stock (Generic, Show, Eq)
   deriving anyclass (GStorable)
 
-newtype DPIDataBuffer = DPIDataBuffer (Ptr DPIDataBuffer)
-  deriving (Show, Eq)
-  deriving newtype (Storable)
+data DPIDataBuffer =
+  AsBoolean Int
+  | AsInt64 Int64
+  | AsUInt64 Word64
+  | AsFloat Float
+  | AsDouble Double
+  | AsString CString
+  | AsBytes DPIBytes
+  | AsTimestamp DPITimeStamp
+  deriving (Show, Eq, Generic)
+
+instance Storable DPIDataBuffer where
+  sizeOf _ = sizeOf (undefined :: DPITimeStamp)
+  alignment _ = 8
+  peek = undefined "don't peek in here!" -- TODO broken type, split into another type for peeking during queries
+
+  poke ptr (AsInt64 intVal) = poke (castPtr ptr) intVal
+  poke ptr (AsUInt64 word64Val) = poke (castPtr ptr) word64Val
+  poke ptr (AsFloat floatVal) = poke (castPtr ptr) floatVal
+  poke ptr (AsDouble doubleVal) = poke (castPtr ptr) doubleVal
+  poke ptr (AsString cStringVal) = poke (castPtr ptr) cStringVal
+  poke ptr (AsBytes dpiBytesVal) = poke (castPtr ptr) dpiBytesVal
+  poke ptr (AsTimestamp dpiTimeStampVal) = poke (castPtr ptr) dpiTimeStampVal
```

Comment:
While poking and peeking works with the auto-derived storable instance using `derive-storable`, the poked data buffer cannot be properly read by ODPIC. I suspect that is because derive-storable does poking in a non-standard way to allow peek-poke interoperability, and thus a poked sum type is not isomorphic to a stored union element.

Reply:
Ayup, turns out that's what it is. According to [their readme](https://github.com/mkloczko/derive-storable/#sum-types),

> Non-sum types have the same memory layout as C structs, **while sum types correspond to tagged unions**

Reply:
Unfortunately that means we can't have a peek instance for this DPIDataBuffer type as it's untagged in memory now.

We will have to use a separate `DPIDataBufferIn` to read values from the buffer. That could be an opaque/abstract pointer, while we use this type to write values to the buffer.